diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
index aeb06b0..bfca98f 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
@@ -79,7 +79,7 @@ static void __iomem *network_interface_setup(struct platform_device *pdev)
 		if (internal_phy == 1) {
 			pr_debug("internal phy\n");
 			writel(0x10110181, PREG_ETH_REG2);
-			writel(0xe489087f, PREG_ETH_REG3);
+			writel(0xe48908ff, PREG_ETH_REG3);
 			pin_ctl = devm_pinctrl_get_select(&pdev->dev,
 				"internal_eth_pins");
 		} else {
