diff --git a/freg_test/Android.mk b/freg_test/Android.mk
new file mode 100755
index 0000000..e3bdd75
--- /dev/null
+++ b/freg_test/Android.mk
@@ -0,0 +1,7 @@
+LOCAL_PATH := $(call my-dir)
+
+include $(CLEAR_VARS)
+LOCAL_MODULE_TAGS := optional
+LOCAL_MODULE := freg
+LOCAL_SRC_FILES := $(call all-subdir-c-files)
+include $(BUILD_EXECUTABLE)
diff --git a/freg_test/freg_test.c b/freg_test/freg_test.c
new file mode 100755
index 0000000..03a6f16
--- /dev/null
+++ b/freg_test/freg_test.c
@@ -0,0 +1,35 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+
+#define FREG_DEVICE_NAME        "/dev/freg"
+
+int main(int argc, char **argv)
+{
+    int fd = -1;
+    int val = 0;
+
+    fd = open(FREG_DEVICE_NAME£¬ O_RDWR);
+    if (fd == -1)
+    {
+        printf("Failed to open device %s.\n"£¬ FREG_DEVICE_NAME);
+        return -1;
+    }
+
+    printf("Read original value:\n");
+    read(fd£¬ &val£¬ sizeof(val));
+    printf("%d.\n\n"£¬ val);
+
+    val = 5;
+    printf("Write value %d to %s.\n\n"£¬ val£¬ FREG_DEVICE_NAME);
+    write(fd£¬ &val£¬ sizeof(val));
+
+    printf("Read the value again:\n");
+    read(fd£¬ &val£¬ sizeof(val));
+    printf("%d.\n\n"£¬ val);
+
+    close(fd);
+
+    return 0
+}
+
